### PR TITLE
Implement customizable summary writer and indicators of attack failure

### LINKED ISSUE #1325: integrate "indicators of attack failure" from Pintor et al 2021
Hi @beat-buesser I don't see an issue for this yet, even though you just mentioned it in your presentation, so I am selfishly raising this issue in the hopes I can track development. If I'm just missing the issue, please feel free to close this.

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Integrate metrics described in https://arxiv.org/abs/2106.09947 that provide quantitative indicators of attack failure.
Implementations here: https://github.com/pralab/IndicatorsOfAttackFailure

This is the specific module:
https://github.com/pralab/IndicatorsOfAttackFailure/blob/master/src/indicators/indicators.py

**Describe alternatives you've considered**
Since you mentioned implementing for certain attacks:

Would it be possible to add these in a way so that user is empowered to apply them with any attack? 

That is, instead of separately implementing additional logic that does all the tracking of metrics, but only for certain attacks, instead give users access to variables they need to compute any of the metrics themselves.

E.g., to compute the `break_point_angle` a user would need `loss`. To get `loss` they would need to pass in a `tensor_board` instance to any attack. 

Currently this is only available for PGD and patch, but it would be easier to implement this for any attack than, say, separately implementing a `compute_breakpoint_angle` method for every single attack.

So with the `tensor_board` approach, these metrics could go in their own module, e.g. `art.attacks.indicators`. 

A user would have to know how to do this, so `art` wouldn't magically do everything for them, but that could be addressed with a Jupyter notebook and/or tutorial in docs.

Maybe this is what you all already have in mind -- just thought I'd offer my 2 cents

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request introduces a new Summary Writer API which allows to collect metrics based on attack and estimator properties as function of iteration step in TensorBoard format. It provides a default summary writer collecting properties of loss gradients, loss values, adversarial patch, and indicators of attack failure. This summary writer will be available in all attacks in `art.attacks` as property `self.summary_writer` and selected attacks including Projected Gradient Descent and Adversarial Patch and Texture have been updated accordingly, more attacks will follow soon. Users can provide customized summary writers to the attacks to collect custom metrics. Introduces a new Union type `ESTIMATOR_TYPE` representing all estimators in ART.

Fixes #1325

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by ambrishrawat — COMMENTED** [review]
Please find comments in the code

File: art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py [hunk 1]
```diff
@@ -458,24 +461,20 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> T
                 for images, target, mask_i in dataset:
                     _ = self._train_step(images=images, target=target, mask=mask_i)
 
+            # Write summary
             if self.summary_writer is not None:  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py:465]: Good point, the super class replace any input type with an instance of `SummaryWriter` or `None`.

File: art/summary_writer.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+            elif hasattr(estimator, "compute_loss"):
+                loss = estimator.compute_loss(x=x, y=y)
+
+                self.summary_writer.add_scalars(
+                    "loss/batch-{}".format(batch_id),
+                    {str(i): v for i, v in enumerate(loss)},
+                    global_step=global_step,
+                )
+
+        # Attack Failure Indicators
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/summary_writer.py:178]: Great suggestion, I will add the reference to Pintor et al. (https://arxiv.org/abs/2106.09947)

File: art/utils.py [hunk 3]
```diff
@@ -203,11 +195,23 @@
...
+        PyTorchObjectDetector,
+        PyTorchFasterRCNN,
+    ]
+
     TENSORFLOWV2_ESTIMATOR_TYPE = Union[  # pylint: disable=C0103
         TensorFlowV2Classifier,
         TensorFlowV2Estimator,
     ]
 
+    ESTIMATOR_TYPE = Union[  # pylint: disable=C0103
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/utils.py:211]: They are related, but I'll also update the PR description.

File: tests/test_summary_writer.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+@pytest.fixture()
+def fix_get_mnist_subset(get_mnist_dataset):
+    (x_train_mnist, y_train_mnist), (x_test_mnist, y_test_mnist) = get_mnist_dataset
+    n_train = 10
+    n_test = 10
+    yield x_train_mnist[:n_train], y_train_mnist[:n_train], x_test_mnist[:n_test], y_test_mnist[:n_test]
+
+
+def test_update(art_warning, fix_get_mnist_subset, image_dl_estimator):
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply tests/test_summary_writer.py]: Good idea, I'll add additional tests.

**Review by ambrishrawat — APPROVED** [review]
Looks good, thanks
